### DISCUSSION TITLE
Maybe poly2tri is not the right delaunay triangulation library for axmol? (can not handle complex polygons easily)

### DISCUSSION THREAD

**Discussion by aismann** [axmolengine/axmol#1384] [General]
see Poly2tri readme:
![image](https://github.com/axmolengine/axmol/assets/8652787/2308bc37-a238-4631-8c58-6051fddb1bf4)

Because of these results on axmol:
![image](https://github.com/axmolengine/axmol/assets/8652787/6703d483-d27a-4a44-a9b3-95c2b02c8b2c)
![image](https://github.com/axmolengine/axmol/assets/8652787/e18c3ddd-92c7-4d9b-a7f8-95973afaa044)

**Comment by aismann** [comment]
@NgVThangBz

can you made some polygon data like above or this one (i belive poly2tri can not handle this kind of polygon sections without add some holes or steiner stuff): 
It will be great if you can check it with poly2tri testbed and tell us the result here.
![image](https://github.com/axmolengine/axmol/assets/8652787/bf76a90e-6914-4aab-bb0b-156a8227212f)

Maybe this is also an usefull test (rotate the same polygon)
![image](https://github.com/axmolengine/axmol/assets/8652787/6658cf39-35c0-42d0-a594-6722c12d5216)

**Reply by NgVThangBz** [reply]
@aismann You need more data to test it right, like the data I provided

**Reply by aismann** [reply]
It will be greate if you test it itself with the testbed of poly2tri and tell use the result here. thanks
(Your data can be used later with cpp-test of course)
Im not so much time at the moment.

Maybe not so big polygons (I belive only the section like above is interesting)

**Reply by NgVThangBz** [reply]
I tested with poly2tri, it won't work correctly if the polylines overlap each other
like that:
![Screenshot 2023-10-12 at 13 07 15](https://github.com/axmolengine/axmol/assets/125545428/52020233-c7c1-4186-a035-1209c2717ab4)

**Reply by aismann** [reply]
A ploy2tri rule is:
![image](https://github.com/axmolengine/axmol/assets/8652787/6dcaaae0-c418-48cd-8642-78a23bfb00cd)

@NgVThangBz 
Can you provide this new data set?
Can you make some data sets like this example:
![image](https://github.com/axmolengine/axmol/assets/8652787/305781fc-968f-4e19-ae40-11478d71a201)

**Reply by NgVThangBz** [reply]
you can check it:
[polyLineData2.txt](https://github.com/axmolengine/axmol/files/12879023/polyLineData2.txt)
[polyLineData.txt](https://github.com/axmolengine/axmol/files/12879025/polyLineData.txt)

**Reply by aismann** [reply]
Have you check it on poly2tri, is it correct on poly2tri? 
![image](https://github.com/axmolengine/axmol/assets/8652787/9dd0f28b-7ab1-492e-a40a-7e475106ffcd)
![image](https://github.com/axmolengine/axmol/assets/8652787/f8774cb0-1ab9-4f08-953e-6b419fcd443b)

**Reply by NgVThangBz** [reply]
@aismann you need data correct on poly2tri?

**Reply by aismann** [reply]
Yes. Both.
But if it on poly2tri wrong there is also wrong on axmol (at the moment)
For the cpp-test I need also "correct filled" complex examples

**Reply by NgVThangBz** [reply]
this is data correct on poly2tri
There are 20 datasets here 
[polyLineData2.txt](https://github.com/axmolengine/axmol/files/12879624/polyLineData2.txt)

**Comment by aismann** [comment]
example with autopolygon
![image](https://github.com/axmolengine/axmol/assets/8652787/600fbef7-89df-44c7-a61d-c66f3f3f5e17)

only one pixel removed:
![image](https://github.com/axmolengine/axmol/assets/8652787/0d18b7cc-a2af-48ec-9f5e-25ce040cc90a)

![image](https://github.com/axmolengine/axmol/assets/8652787/b74da1ad-5ba0-43a7-ba9f-cc5b8b4e8c2f)

Removed some pixel
![image](https://github.com/axmolengine/axmol/assets/8652787/0f18c4a0-ac7a-4999-ac4d-dd01cf495300)

Seems the distance have to be "4 pixel" between two "lines"

**Comment by aismann** [comment]
Fazit:
poly2tri works not for all "complex" Polygons, but its better as the "cocos2dx 4.0" implementation:
![image](https://github.com/axmolengine/axmol/assets/8652787/5f44b29f-eab8-48bd-8d5e-084cc2ca7042)
![image](https://github.com/axmolengine/axmol/assets/8652787/35ee06fe-89d4-4b17-97d0-fa4587e1d96f)

Cocos2dx 4.0/Axmol (old):
![image](https://github.com/axmolengine/axmol/assets/8652787/1bc684e1-a323-41f6-9f28-3bd30679d13a)

Axmol (new)
![image](https://github.com/axmolengine/axmol/assets/8652787/5409d3af-ac99-44d5-b3a4-0410e2710a53)

Only for me (autopolygon data set?):
![image](https://github.com/axmolengine/axmol/assets/8652787/8e837c0d-73ab-4994-977d-7e9efafbe5ce)

**Comment by aismann** [comment]
@halx99 
Where can I put a new math function:   ` bool isConvex(const Vec2* verts, int count)`?
Its only used on DrawNopde.cpp

**Comment by aismann** [comment]
I close it now
![image](https://github.com/axmolengine/axmol/assets/8652787/dfdb4108-c1e5-4e9b-b0ec-4a22c49700f0)
